test/bio_memleak_test.c

```diff
@@ -221,6 +221,68 @@ static int test_bio_nonclear_rst(void)
     return ok;
 }
 
+static int error_callback_fired;
+long BIO_error_callback(BIO *bio, int cmd, const char *argp,
+                        int argi, long argl, long ret) {
+
+    if ((cmd & BIO_CB_READ) && (cmd & BIO_CB_RETURN)) {
+        error_callback_fired = 1;
+        ret = 0;  // fail for read operations to simulate error in input BIO
+    }
+    return ret;
+}
+
+// Checks i2d_ASN1_bio_stream() is freeing all memory when input BIO ends unexpectedly.
+static int test_bio_i2d_ASN1_mime(void)
+{
+    int ok = 0;
+    BIO *bio = NULL, *out = NULL;
+    BUF_MEM bufmem;
+    static const char str[] = "BIO mime test\n";
+    PKCS7 *p7 = NULL;
+    error_callback_fired = 0;
+
+    bio = BIO_new(BIO_s_mem());
+    if (!TEST_ptr(bio))
+        goto finish;
+
+    bufmem.length = sizeof(str);
+    bufmem.data = (char *) str;
+    bufmem.max = bufmem.length;
+    BIO_set_mem_buf(bio, &bufmem, BIO_NOCLOSE);
+    BIO_set_flags(bio, BIO_FLAGS_MEM_RDONLY);
+    BIO_set_callback(bio, BIO_error_callback);
+
+    out = BIO_new(BIO_s_mem());
+    if (!TEST_ptr(out))
+        goto finish;
+
+    p7 = PKCS7_new();
+    if (!TEST_ptr(p7))
+        goto finish;
+    if (!PKCS7_set_type(p7, NID_pkcs7_data))
+        goto finish;
+
+    // the call succeeds even if the input stream ends unexpectedly as
+    // there is no handling for this case in SMIME_crlf_copy()
+    if (!i2d_ASN1_bio_stream(out, (ASN1_VALUE*) p7, bio,
+                             SMIME_STREAM | SMIME_BINARY,
+                             ASN1_ITEM_rptr(PKCS7))) {
+        goto finish;
+    }
+
+    if (!TEST_int_eq(1, error_callback_fired))
+        goto finish;
+
+    ok = 1;
+
+    finish:
```

Comment:
nit: just one space indentation on this goto label

Reply:
Fixed, thanks